From c89808bbb434f4a44057b33e0e28530ab08ad016 Mon Sep 17 00:00:00 2001
From: Matthias Clasen <mclasen@redhat.com>
Date: Fri, 8 Nov 2024 10:57:46 -0500
Subject: [PATCH 6/7] win32: Require dwrite_3.h

Requiring a new enough SDK simplifies our code.

(cherry picked from commit aa3df5b0801db377045125190b5977ffb4252ad7)
---
 meson.build                              |  6 ++---
 pango/pangowin32-dwrite-fontmap.cpp      | 30 ++++--------------------
 pango/pangowin32-dwrite-utils-legacy.cpp | 12 +---------
 3 files changed, 7 insertions(+), 41 deletions(-)

diff --git a/meson.build b/meson.build
index e474fc04..fc5324d5 100644
--- a/meson.build
+++ b/meson.build
@@ -350,10 +350,8 @@ if host_system == 'darwin'
 endif
 
 # Check for dwrite_3.h (from more recent Windows SDK or mingw-w64)
-if host_system == 'windows'
-  if cpp.has_header('dwrite_3.h')
-    pango_conf.set('HAVE_DWRITE_3_H', 1)
-  endif
+if host_system == 'windows' and not cpp.has_header('dwrite_3.h')
+  error('Windows backend enabled but dwrite_3.h not found.')
 endif
 
 # Enable cairo-ft with FreeType and FontConfig support if
diff --git a/pango/pangowin32-dwrite-fontmap.cpp b/pango/pangowin32-dwrite-fontmap.cpp
index 810a58d0..9aed0f71 100644
--- a/pango/pangowin32-dwrite-fontmap.cpp
+++ b/pango/pangowin32-dwrite-fontmap.cpp
@@ -23,9 +23,6 @@
 #include "config.h"
 
 #include <initguid.h>
-
-#ifdef HAVE_DWRITE_3_H
-/* we need dwrite_3.h for IDWriteFactory[3|5] */
 #include <dwrite_3.h>
 
 # ifdef _MSC_VER
@@ -42,16 +39,6 @@ struct _PangoWin32DWriteFontSetBuilder
   IDWriteFontSetBuilder *font_set_builder;
 };
 
-#else
-/* stub, for simplicity reasons, if we don't have IDWriteFactory[3|5] */
-# define IDWriteFactory3 IUnknown
-# define IDWriteFactory5 IUnknown
-# define IDWriteFontSet IUnknown
-# define IDWriteFontSetBuilder IUnknown
-# define IDWriteFontSetBuilder1 IUnknown
-#endif
-#include <dwrite_1.h>
-
 #ifdef STRICT
 #undef STRICT
 #endif
@@ -83,12 +70,11 @@ pango_win32_init_direct_write (void)
   gboolean have_idwritefactory3 = FALSE;
   gboolean have_idwritefactory5 = FALSE;
 
-#ifdef HAVE_DWRITE_3_H
   /* Try to create a IDWriteFactory3 first, which is available on Windows 10+ */
   hr = DWriteCreateFactory (DWRITE_FACTORY_TYPE_SHARED,
                             UUID_OF_IDWriteFactory3,
                             reinterpret_cast<IUnknown**> (&factory3));
-  if (SUCCEEDED(hr))
+  if (SUCCEEDED (hr))
     {
       /*
        * Try to acquire a IDWriteFactory5 object from the IDWriteFactory3 object,
@@ -104,7 +90,6 @@ pango_win32_init_direct_write (void)
                                      reinterpret_cast<void**> (&factory));
     }
   else
-#endif
     hr = DWriteCreateFactory (DWRITE_FACTORY_TYPE_SHARED,
                               UUID_OF_IDWriteFactory,
                               reinterpret_cast<IUnknown**> (&factory));
@@ -329,7 +314,6 @@ pango_win32_dwrite_font_map_populate (PangoWin32FontMap *map)
 
   pango_win32_dwrite_font_map_populate_with_collection (map, sys_collection);
 
-#ifdef HAVE_DWRITE_3_H
   /* the following code requires items from dwrite_3.h */
   if (dwrite_items->have_idwritefactory5 || dwrite_items->have_idwritefactory3)
     {
@@ -378,16 +362,14 @@ pango_win32_dwrite_font_map_populate (PangoWin32FontMap *map)
 
       fontset->Release ();
     }
-  else
-#endif /* HAVE_DWRITE_3_H */
-  if (map->custom_fonts_legacy != NULL &&
-      map->custom_fonts_legacy->font_collection_temp != NULL)
+  else if (map->custom_fonts_legacy != NULL &&
+           map->custom_fonts_legacy->font_collection_temp != NULL)
     {
       IDWriteFontCollection *collection = map->custom_fonts_legacy->font_collection_temp;
 
       pango_win32_dwrite_font_map_populate_with_collection (map, collection);
       collection->Release ();
-	  map->custom_fonts_legacy->font_collection_temp = NULL;
+          map->custom_fonts_legacy->font_collection_temp = NULL;
     }
 
   sys_collection->Release ();
@@ -761,7 +743,6 @@ pango_win32_font_create_hb_face_dwrite (PangoWin32Font *font)
 }
 #endif
 
-#ifdef HAVE_DWRITE_3_H
 /* the following items require items from dwrite_3.h */
 static gboolean
 add_custom_font_factory5 (PangoFontMap     *font_map,
@@ -975,8 +956,6 @@ pango_win32_dwrite_add_font_file (PangoFontMap *font_map,
 
   dwrite_items = pango_win32_get_direct_write_items ();
 
-#ifdef HAVE_DWRITE_3_H
-  /* we don't support custom font loading yet for pre-Windows 10 */
   if (dwrite_items->have_idwritefactory5)
     succeeded = add_custom_font_factory5 (font_map,
                                           dwrite_items->dwrite_factory5,
@@ -988,7 +967,6 @@ pango_win32_dwrite_add_font_file (PangoFontMap *font_map,
                                           font_file_path,
                                           error);
   else
-#endif
     succeeded = add_custom_font_legacy (font_map,
                                         dwrite_items->dwrite_factory,
                                         font_file_path,
diff --git a/pango/pangowin32-dwrite-utils-legacy.cpp b/pango/pangowin32-dwrite-utils-legacy.cpp
index 519a573a..a8c9f271 100644
--- a/pango/pangowin32-dwrite-utils-legacy.cpp
+++ b/pango/pangowin32-dwrite-utils-legacy.cpp
@@ -30,17 +30,7 @@
 #include <string>
 #include <vector>
 
-/* stub, for simplicity reasons, if we don't have IDWriteFactory[3|5] */
-#ifdef HAVE_DWRITE_3_H
-# include <dwrite_3.h>
-#else
-# include <dwrite_1.h>
-# define IDWriteFactory3 IUnknown
-# define IDWriteFactory5 IUnknown
-# define IDWriteFontSet IUnknown
-# define IDWriteFontSetBuilder IUnknown
-# define IDWriteFontSetBuilder1 IUnknown
-#endif
+#include <dwrite_3.h>
 
 #include "pangowin32-private.hpp"
 
-- 
2.46.0

